src/transport/quinn.rs

```diff
@@ -1,78 +1,302 @@
 //! QUIC channel implementation based on quinn
 use crate::{LocalAddr, RpcMessage};
+use futures::channel::oneshot;
+use futures::future::BoxFuture;
 use futures::{Future, FutureExt, Sink, SinkExt, Stream, StreamExt};
 use pin_project::pin_project;
 use serde::{de::DeserializeOwned, Serialize};
 use std::net::SocketAddr;
+use std::sync::Arc;
+use std::task::{self, ready, Poll};
 use std::{fmt, io, marker::PhantomData, pin::Pin, result};
 use tokio_serde::{formats::SymmetricalBincode, SymmetricallyFramed};
 use tokio_util::codec::{FramedRead, FramedWrite, LengthDelimitedCodec};
 
 type Socket<In, Out> = (SendSink<Out>, RecvStream<In>);
 
+#[derive(Debug)]
+struct ServerChannelInner {
+    endpoint: Option<quinn::Endpoint>,
+    task: Option<tokio::task::JoinHandle<()>>,
+    local_addr: [LocalAddr; 1],
+    receiver: flume::Receiver<SocketInner>,
+}
+
+impl Drop for ServerChannelInner {
+    fn drop(&mut self) {
+        if let Some(endpoint) = self.endpoint.take() {
+            endpoint.close(0u32.into(), b"server channel dropped");
+        }
+        if let Some(task) = self.task.take() {
+            task.abort()
```

Comment:
this is the brute force drop variant that just drops the task immediately. I think once you get to drop you should clean things up quickly, but there could be an additional fn that does a more graceful shutdown.